magma/topological_sort.py

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python3
+
+################################################
+#   Libraries
+################################################
+from copy import deepcopy
+
+################################################
+#   Functions
+################################################
+
+def check_presence_of_key(list_of_dicts, key_to_check=None):
+    """
+    Takes in a list of dictionaries and a list of keys, checks that those keys 
+    are present within every dict in this list/array.
+
+    :param list_of_dicts: dictionaries to check
+    :type input_dict: list[dict]
+    :param key_to_check: key to check
+    :type key_to_check: str
+    :return: True, if the specified key is present in each dict, or there is no 
+                                key to check, else False
+    :rtype: bool
+    """
+    #TODO: make the next three commands its own helper function? I repeat variations
+    # several times -- on this note, look up how to have flexible argument
+    if key_to_check is None:
+        return True
+
+    if not all((key_to_check in dictionary) for dictionary in list_of_dicts):
+        return False
+
+    return True
+
+#TODO: could make this more general...
+def set_dependency_list_values(list_of_dicts, name_of_step_key, name_of_dependencies_key):
+    """
+    Checks for dependency key within each dictionary in list_of_dicts.
+    If not present, add that key and set value as empty list.
+    Else, remove duplicates and self-dependencies.
+
+    :param list_of_dicts: list of dictionaries that should hold a dependency list. Each
+                                dictionary corresponds to a step, and the list represents
+                                a "list of steps" with dependencies between them.
+    :type list_of_dicts: list[dict]
+    :param name_of_step_key: name of the key corresponding to the step's name –
+                                    i.e. attribute referred to by dependency values
+    :type name_of_step_key: str
+    :param name_of_dependencies_key: name of the key corresponding to the dependencies list
+    :type name_of_dependencies_key: str
+    :return: a copy of list_of_dicts with appropriate dependency lists set
+    :rtype: list[dict]
+    """
+
+    list_of_dicts_copy = deepcopy(list_of_dicts)
+    # iterate through list of dicts and set dependencies key-value pair
+    for dictionary in list_of_dicts_copy:
+        # add empty dependency list if not present
+        if not name_of_dependencies_key in dictionary:
+            dictionary[name_of_dependencies_key] = []
+            #TODO: do some renaming of this function to follow pattern of obj vs dict key setting?
+
+        # get rid of duplicates
+        # I choose this method for generalization, in the case that dependencies is
+        # a list of dictionaries, which are an unhashable type
```

Comment:
The only difference I would make with the assumption that dependencies are not of unhashable (they are, indeed, strings) is that I would get rid of duplicates using the set() method instead of the "non_duplicated_dependencies" array creation (see your next comment). If I do the set() method, it would work with this schema. If there were ever a case where this code were passed a list of dictionaries (or other unhashable type) as the dependencies, it would throw a TypeError. Might switch it to the set() method for simplicity.